sdk/opa.go

```diff
@@ -0,0 +1,234 @@
+
+	for _, opt := range opts {
+		if err := opt(opa); err != nil {
+			return nil, err
+		}
+	}
+
+	id, err := uuid.New(rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	var pluginInitializers []func(*plugins.Manager)
+
+	if opa.paths != nil && len(opa.paths) > 0 {
+		loaded, err := initload.LoadPaths(opa.paths, opa.filter, opa.bundleMode, nil, false)
+		if err != nil {
+			return nil, fmt.Errorf("load error %v", err)
+		}
+		pluginInitializers = append(pluginInitializers, plugins.InitBundles(loaded.Bundles))
+		pluginInitializers = append(pluginInitializers, plugins.InitFiles(loaded.Files))
+	}
+
+	if opa.store == nil {
+		opa.store = inmem.New()
+	}
+
+	opa.manager, err = plugins.New(opa.configBytes, id, opa.store, pluginInitializers...)
+	if err != nil {
+		return nil, err
+	}
+
+	discovery, err := discovery.New(opa.manager)
+	if err != nil {
+		return nil, err
+	}
+
+	opa.manager.RegisterCompilerTrigger(opa.compilerUpdated)
+	opa.manager.Register("discovery", discovery)
+
+	opa.interQueryBuiltinCache = iCache.NewInterQueryCache(opa.manager.InterQueryBuiltinCacheConfig())
+
+	opa.preparedQueries = map[string]*rego.PreparedEvalQuery{}
+	opa.preparedQueriesDoOnce = map[string]*sync.Once{}
+
+	return opa, nil
+}
+
+// Start asynchronously starts the policy engine's plugins that download
+// policies, report status, etc.
+func (opa *OPA) Start(ctx context.Context) error {
+	return opa.manager.Start(ctx)
+}
+
+// Query sends a query to the OPA instance and returns the decision
+func (opa *OPA) Query(ctx context.Context, decisionPath string, input interface{}, opts ...func(*rego.Rego)) (interface{}, error) {
+	m := metrics.New()
+	var decisionID string
+	var decision interface{}
+
+	err := storage.Txn(ctx, opa.manager.Store, storage.TransactionParams{}, func(txn storage.Transaction) error {
+		var err error
+		decisionID, err = uuid.New(rand.Reader)
+		if err != nil {
+			return err
+		}
+
+		if _, ok := opa.preparedQueriesDoOnce[decisionPath]; !ok {
+			opa.preparedQueriesDoOnce[decisionPath] = &sync.Once{}
+		}
+
+		var pq rego.PreparedEvalQuery
+		opa.preparedQueriesDoOnce[decisionPath].Do(func() {
+			opts = append(opts,
+				rego.Metrics(m),
+				rego.Query(decisionPath),
+				rego.Compiler(opa.manager.GetCompiler()),
+				rego.Store(opa.manager.Store),
+				rego.Transaction(txn),
+				rego.Runtime(opa.manager.Info),
+			)
+			r := rego.New(opts...)
+			pq, err = r.PrepareForEval(context.Background())
+			opa.preparedQueries[decisionPath] = &pq
+		})
+
+		rs, err := opa.preparedQueries[decisionPath].Eval(
+			ctx,
+			rego.EvalInput(input),
+			rego.EvalTransaction(txn),
+			rego.EvalMetrics(m),
+			rego.EvalInterQueryBuiltinCache(opa.interQueryBuiltinCache),
+		)
+
+		if err != nil {
+			return err
+		} else if len(rs) == 0 {
+			return fmt.Errorf("undefined decision")
+		} else {
+			decision = rs[0].Expressions[0].Value
+		}
+
+		if logger := logs.Lookup(opa.manager); logger != nil {
+			bundles, err := opa.bundles(ctx, txn)
```

Comment:
❓ If this `err` is _not_ `nil`, will this have an affect? Or rather, the err that's logged below, https://github.com/open-policy-agent/opa/pull/3186/files#diff-9402e37dbcd2155047270666944796167deea341d9c6288dda653ca82b30ce87R193, isn't it meant to be the `err` of https://github.com/open-policy-agent/opa/pull/3186/files#diff-9402e37dbcd2155047270666944796167deea341d9c6288dda653ca82b30ce87R168, not this one?

Reply:
Yeah that makes sense. I believe this is due to `opa.bundles` using `panic` before so when this was changed to return an error instead this happened. I'll fix!